django/db/migrations/operations/models.py

```diff
@@ -802,6 +805,7 @@ def state_forwards(self, app_label, state):
         model_state = state.models[app_label, self.model_name_lower]
         indexes = model_state.options[self.option_name]
         model_state.options[self.option_name] = [idx for idx in indexes if idx.name != self.name]
+        state.reload_model(app_label, self.model_name_lower, delay=True)
```

Comment:
The test should check that the model class on an old and a new `ProjectState`s changed: here are appropriate test methods for `AddIndex` and `RemoveIndex` based on 04810ea43373b8810be9bc6c47ea7aacfa178c53:
```diff
diff --git a/tests/migrations/test_operations.py b/tests/migrations/test_operations.py
index 22879d884e..a6f713e321 100644
--- a/tests/migrations/test_operations.py
+++ b/tests/migrations/test_operations.py
@@ -1490,17 +1490,27 @@ class OperationTests(OperationTestBase):
         self.assertIndexExists("test_rmin_pony", ["pink", "weight"])
 
     def test_add_index_state_forwards(self):
-        project_state = self.set_up_test_model('test_adin')
-        index = models.Index(fields=['pink'], name='test_adin_pony_pink_idx')
+        project_state = self.set_up_test_model('test_adinsf')
+        index = models.Index(fields=['pink'], name='test_adinsf_pony_pink_idx')
+        old_model = project_state.apps.get_model('test_adinsf', 'Pony')
+        new_state = project_state.clone()
+
         operation = migrations.AddIndex('Pony', index)
+        operation.state_forwards('test_adinsf', new_state)
+        new_model = new_state.apps.get_model('test_adinsf', 'Pony')
+        self.assertIsNot(old_model, new_model)
+
+    def test_remove_index_state_forwards(self):
+        project_state = self.set_up_test_model('test_rminsf')
+        index = models.Index(fields=['pink'], name='test_rminsf_pony_pink_idx')
+        migrations.AddIndex('Pony', index).state_forwards('test_rminsf', project_state)
+        old_model = project_state.apps.get_model('test_rminsf', 'Pony')
         new_state = project_state.clone()
-        operation.state_forwards('test_adin', new_state)
 
-        before_indexes = project_state.models['test_adin', 'pony'].options['indexes']
-        self.assertEqual(before_indexes, [])
-        after_indexes = new_state.models['test_adin', 'pony'].options['indexes']
-        self.assertEqual(after_indexes, [index])
-        self.assertIsNot(after_indexes[0], index)
+        operation = migrations.RemoveIndex('Pony', 'test_rminsf_pony_pink_idx')
+        operation.state_forwards('test_rminsf', new_state)
+        new_model = new_state.apps.get_model('test_rminsf', 'Pony')
+        self.assertIsNot(old_model, new_model)
 
     def test_alter_field_with_index(self):
         """
```